pwndbg/emu/emulator.py

```diff
@@ -667,8 +686,21 @@ def emu_stop(self, *a, **kw):
 
     def emulate_with_hook(self, hook, count=512) -> None:
         ident = self.hook_add(U.UC_HOOK_CODE, hook)
+
+        pc: int = self.pc
+        if thumb_bit := self.read_thumb_bit():
+            # Unicorn appears to disregard the UC_MODE_THUMB mode passed into the constructor, and instead
+            # determines Thumb mode based on the PC that is passed to the `emu_start` function
+            # https://github.com/unicorn-engine/unicorn/issues/391
+            #
+            # Because we single-step the emulator, we always have to read the Thumb bit from the emulator
+            # and set the least significant bit of the PC to 1 if the bit is 1 in order to enable Thumb mode
+            # for the execution of the next instruction. If this `emulate_with_hook` executes multiple instructions
+            # which have Thumb mode transitions, Unicorn will internally handle them.
+            pc |= thumb_bit
```

Comment:
Note: we could probably drop the `if` and always do the bit-or operation? I bet we can't because this may return a `None`? We could in theory always return `0` instead... heh.

Reply:
Done!